File: packages/adapter-langchain/src/tools.ts
import { StructuredTool } from '@langchain/core/tools'
import { z } from 'zod'
import { PlaywrightRunner, capturePlaywrightSnapshot } from '@quokka/runner-playwright'
import { RecipeV2Schema } from '@quokka/shared'
import type { LLMProvider } from '@quokka/core'
import { chromium, type Browser, type Page } from 'playwright'

// ---------------------------------------------------------------------------
// Shared helpers
// ---------------------------------------------------------------------------

interface QuokkaToolConfig {
  /** LLM provider for observe/plan tools */
  provider?: LLMProvider
  /** Browser launch options */
  headless?: boolean
  /** Per-action timeout in ms (default: 30000) */
  timeout?: number
  /** Viewport size */
  viewport?: { width: number; height: number }
}

async function launchBrowser(config: QuokkaToolConfig): Promise<{ browser: Browser; page: Page }> {
  const browser = await chromium.launch({ headless: config.headless ?? true })
  const page = await browser.newPage({
    viewport: config.viewport ?? { width: 1280, height: 720 },
  })
  if (config.timeout) page.setDefaultTimeout(config.timeout)
  return { browser, page }
}

// ---------------------------------------------------------------------------
// QuokkaExecuteTool
// ---------------------------------------------------------------------------

export class QuokkaExecuteTool extends StructuredTool {
  name = 'quokka_execute'
  description =
    'Execute a browser automation task. Provide a natural-language intent ' +
    'and a starting URL. Returns the execution result including status and ' +
    'number of steps executed.'

  schema = z.object({
    intent: z.string().describe('Natural-language description of what to do in the browser'),
    url: z.string().url().describe('Starting URL to navigate to'),
    variables: z
      .record(z.string(), z.string())
      .optional()
      .describe('Key-value variables to inject into the recipe'),
  })

  private config: QuokkaToolConfig

  constructor(config: QuokkaToolConfig = {}) {
    super()
    this.config = config
  }

  async _call(input: z.infer<typeof this.schema>): Promise<string> {
    const recipe = RecipeV2Schema.parse({
      version: '2.0',
      id: `langchain-exec-${Date.now()}`,
      name: 'LangChain Execute',
      intent: input.intent,
      steps: [
        {
          id: 'step-1',
          type: 'action',
          intent: input.intent,
          likelyNavigates: true,
        },
      ],
      hosts: [new URL(input.url).hostname],
    })

    const runner = new PlaywrightRunner({
      headless: this.config.headless ?? true,
      timeout: this.config.timeout,
      viewport: this.config.viewport,
    })

    try {
      const result = await runner.run(
        recipe,
        { ...input.variables, __startUrl: input.url },
        this.config.provider!,
      )
      return JSON.stringify(result, null, 2)
    } finally {
      await runner.close()
    }
  }
}

// ---------------------------------------------------------------------------
// QuokkaObserveTool
// ---------------------------------------------------------------------------

export class QuokkaObserveTool extends StructuredTool {
  name = 'quokka_observe'
  description =
    'Navigate to a URL, capture a DOM snapshot of the page, and answer a ' +
    'question about what is visible. Returns the LLM answer based on the ' +
    'live page content.'

  schema = z.object({
    url: z.string().url().describe('URL to navigate to and observe'),
    question: z.string().describe('Question to answer about the page content'),
  })

  private config: QuokkaToolConfig

  constructor(config: QuokkaToolConfig = {}) {
    super()
    this.config = config
    if (!config.provider) {
      throw new Error('QuokkaObserveTool requires an LLMProvider in config')
    }
  }

  async _call(input: z.infer<typeof this.schema>): Promise<string> {
    const { browser, page } = await launchBrowser(this.config)

    try {
      await page.goto(input.url, { waitUntil: 'domcontentloaded' })
      const snapshot = await capturePlaywrightSnapshot(page)

      const answer = await this.config.provider!.complete([
        {
          role: 'system',
          content:
            'You are analyzing a web page. Answer the user question based on ' +
            'the following DOM snapshot.\n\n' +
            `URL: ${snapshot.url}\nTitle: ${snapshot.title}\n\n` +
            'Accessibility tree:\n' +
            snapshot.accessibilityTree
              .map((n) => `[${n.role}] "${n.name}" (${n.selector})`)
              .join('\n'),
        },
        { role: 'user', content: input.question },
      ])

      return JSON.stringify({ url: snapshot.url, title: snapshot.title, answer })
    } finally {
      await page.close()
      await browser.close()
    }
  }
}

// ---------------------------------------------------------------------------
// QuokkaPlanTool
// ---------------------------------------------------------------------------

export class QuokkaPlanTool extends StructuredTool {
  name = 'quokka_plan'
  description =
    'Decompose a high-level browser automation goal into a sequence of ' +
    'concrete steps. Returns a RecipeV2 JSON with ordered action steps.'

  schema = z.object({
    goal: z.string().describe('High-level goal to decompose into browser steps'),
    url: z.string().url().describe('Starting URL for the task'),
    maxSteps: z
      .number()
      .int()
      .min(1)
      .max(20)
      .optional()
      .default(10)
      .describe('Maximum number of steps to generate (default: 10)'),
  })

  private config: QuokkaToolConfig

  constructor(config: QuokkaToolConfig = {}) {
    super()
    this.config = config
    if (!config.provider) {
      throw new Error('QuokkaPlanTool requires an LLMProvider in config')
    }
  }

  async _call(input: z.infer<typeof this.schema>): Promise<string> {
    const provider = this.config.provider!

    const response = await provider.completeJSON<{
      steps: Array<{ intent: string; context_hint?: string; likelyNavigates: boolean }>
    }>(
      [
        {
          role: 'system',
          content:
            'You are a browser automation planner. Given a goal and starting URL, ' +
            'decompose it into concrete sequential steps. Each step should be a ' +
            'single user-visible action (click, type, navigate, etc.).',
        },
        {
          role: 'user',
          content:
            `Goal: ${input.goal}\nStarting URL: ${input.url}\n` +
            `Generate up to ${input.maxSteps} steps.`,
        },
      ],
      {
        type: 'object',
        properties: {
          steps: {
            type: 'array',
            items: {
              type: 'object',
              properties: {
                intent: { type: 'string' },
                context_hint: { type: 'string' },
                likelyNavigates: { type: 'boolean' },
              },
              required: ['intent', 'likelyNavigates'],
            },
          },
        },
        required: ['steps'],
      },
    )

    const recipe = RecipeV2Schema.parse({
      version: '2.0',
      id: `langchain-plan-${Date.now()}`,
      name: `Plan: ${input.goal.slice(0, 60)}`,
      description: input.goal,
      intent: input.goal,
      steps: response.steps.map((s, i) => ({
        id: `step-${i + 1}`,
        type: 'action' as const,
        intent: s.intent,
        context_hint: s.context_hint,
        likelyNavigates: s.likelyNavigates,
      })),
      hosts: [new URL(input.url).hostname],
    })

    return JSON.stringify(recipe, null, 2)
  }
}

File: packages/adapter-crewai/src/quokka_crewai/tools.py
"""Quokka browser automation tools for CrewAI.

Each tool connects to the Quokka MCP server via subprocess stdio transport
and delegates to the corresponding MCP tool (quokka_execute, quokka_observe,
quokka_plan).
"""

from __future__ import annotations

import asyncio
import json
from typing import Any, Optional, Type

from crewai.tools import BaseTool
from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client
from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# MCP connection helper
# ---------------------------------------------------------------------------

_DEFAULT_SERVER_PARAMS = StdioServerParameters(
    command="npx",
    args=["quokka-mcp"],
)


async def _call_mcp_tool(
    tool_name: str,
    arguments: dict[str, Any],
    server_params: StdioServerParameters | None = None,
) -> str:
    """Open a short-lived MCP session and call a single tool."""
    params = server_params or _DEFAULT_SERVER_PARAMS

    async with stdio_client(params) as (read_stream, write_stream):
        async with ClientSession(read_stream, write_stream) as session:
            await session.initialize()
            result = await session.call_tool(tool_name, arguments)
            # MCP returns content as a list of content blocks
            parts = []
            for block in result.content:
                if hasattr(block, "text"):
                    parts.append(block.text)
            return "\n".join(parts) if parts else json.dumps({"status": "ok"})


def _run_async(coro: Any) -> Any:
    """Run an async coroutine from sync context."""
    try:
        loop = asyncio.get_running_loop()
    except RuntimeError:
        loop = None

    if loop and loop.is_running():
        # We're inside an existing event loop (e.g. Jupyter, CrewAI async)
        import concurrent.futures
        with concurrent.futures.ThreadPoolExecutor() as pool:
            return pool.submit(asyncio.run, coro).result()
    else:
        return asyncio.run(coro)


# ---------------------------------------------------------------------------
# Input schemas
# ---------------------------------------------------------------------------

class ExecuteInput(BaseModel):
    intent: str = Field(description="Natural-language description of what to do in the browser")
    url: str = Field(description="Starting URL to navigate to")
    variables: Optional[dict[str, str]] = Field(
        default=None,
        description="Key-value variables to inject into the recipe",
    )


class ObserveInput(BaseModel):
    url: str = Field(description="URL to navigate to and observe")
    question: str = Field(description="Question to answer about the page content")


class PlanInput(BaseModel):
    goal: str = Field(description="High-level goal to decompose into browser steps")
    url: str = Field(description="Starting URL for the task")
    max_steps: int = Field(
        default=10,
        ge=1,
        le=20,
        description="Maximum number of steps to generate (default: 10)",
    )


# ---------------------------------------------------------------------------
# CrewAI Tools
# ---------------------------------------------------------------------------

class QuokkaExecuteTool(BaseTool):
    """Execute a browser automation task via Quokka."""

    name: str = "quokka_execute"
    description: str = (
        "Execute a browser automation task. Provide a natural-language intent "
        "and a starting URL. Returns the execution result including status and "
        "number of steps executed."
    )
    args_schema: Type[BaseModel] = ExecuteInput
    server_params: Optional[StdioServerParameters] = None

    model_config = {"arbitrary_types_allowed": True}

    def _run(
        self,
        intent: str,
        url: str,
        variables: dict[str, str] | None = None,
    ) -> str:
        args: dict[str, Any] = {"intent": intent, "url": url}
        if variables:
            args["variables"] = variables
        return _run_async(_call_mcp_tool("quokka_execute", args, self.server_params))


class QuokkaObserveTool(BaseTool):
    """Navigate to a URL and answer a question about the page content."""

    name: str = "quokka_observe"
    description: str = (
        "Navigate to a URL, capture a DOM snapshot of the page, and answer a "
        "question about what is visible. Returns the LLM answer based on the "
        "live page content."
    )
    args_schema: Type[BaseModel] = ObserveInput
    server_params: Optional[StdioServerParameters] = None

    model_config = {"arbitrary_types_allowed": True}

    def _run(self, url: str, question: str) -> str:
        return _run_async(
            _call_mcp_tool(
                "quokka_observe",
                {"url": url, "question": question},
                self.server_params,
            )
        )


class QuokkaPlanTool(BaseTool):
    """Decompose a high-level goal into concrete browser automation steps."""

    name: str = "quokka_plan"
    description: str = (
        "Decompose a high-level browser automation goal into a sequence of "
        "concrete steps. Returns a recipe JSON with ordered action steps."
    )
    args_schema: Type[BaseModel] = PlanInput
    server_params: Optional[StdioServerParameters] = None

    model_config = {"arbitrary_types_allowed": True}

    def _run(self, goal: str, url: str, max_steps: int = 10) -> str:
        return _run_async(
            _call_mcp_tool(
                "quokka_plan",
                {"goal": goal, "url": url, "maxSteps": max_steps},
                self.server_params,
            )
        )
